### DISCUSSION TITLE
Export stage's graphics to livestreaming softwares like OBS, vMix?

### DISCUSSION THREAD

**Discussion by tthtth123** [airsdk/Adobe-Runtime-Support#1783] [Ideas]
How do you think about a library to export AS3 animations to streaming softwares like OBS, vMix...? I don't want to use the "chroma key" technique because its limitations.

**Comment by ajwfrost** [comment]
Wow I would love to have the possibility of SWF apps being able to display within OBS or similar. We actually did a project many years ago for a media company so that they can have SWF overlays (so like when you're watching TV and an animated channel logo appears in the top corner etc..)

My one big concern:
_OBS Studio is an open source program licensed under the [GPLv2](https://github.com/obsproject/obs-studio/blob/master/COPYING). That means that if you create a plugin for OBS Studio or a modified version of OBS Studio and distribute it in binary form, you must give users a way to access the source code of your binary._

So you may need to continue with window capture -> could you set up the AIR application to be fully transparent though, would that help?

**Reply by tthtth123** [reply]
No, OBS can't capture a transparent window keeping transparent graphics!

**Reply by ajwfrost** [reply]
Okay .. so let me check with our OSS lawyers, but I've just seen this:
https://www.gnu.org/licenses/gpl-faq.en.html#MereAggregation

Which implies that it may be possible for us (or anyone) to create a plug-in for OBS (open source/GPL-license) if it worked with AIR via a simple/standard IPC mechanism ... and actually the obvious thing, at least on Windows, would be to have this as a "virtual webcam".

Some information (adding the link here for reference): https://stackoverflow.com/questions/33693131/how-to-create-virtual-webcam-in-windows-10
and it seems to work best if we used DirectShow: https://docs.microsoft.com/en-us/windows/win32/directshow/csource
The challenge may be around alpha channels: DirectShow supports video with an alpha channel: https://docs.microsoft.com/en-us/windows/win32/directshow/alpha-blending 
but this would be uncompressed video (well, that makes it easier for us) and I'm not 100% sure whether that would be supported by OBS etc....

Interesting one! We will have to look into this further :-)

**Reply by tthtth123** [reply]
Hi ajwfrost,
Thanks for your ideas. I think the thing is not only the plug-in for OBS but also an library for AS3 (that can capture transparent graphics well).
For OBS graphic input, I found the NDI technique.
This is how I've tried:
I have an AIR app running graphics. The app use _**flash.display.BitmapData.draw()/drawWithQuality()**_ method in AS3 to render the graphic (frame-by-frame) on the stage as BitmapData, transfer it into FFMPEG.exe, and FFMPEG.exe will give it to NDI output.
1. But the result from the "draw()" method makes the animation playing slowly. There's a delay time and sometimes it's not good. I want the speed as "capture window" in OBS (but the problem is capturing doesn't support transperency!).
Moreover, the "draw()" method can not render 3D objects created on the stage (I am using Adobe Animate and create those 3D objects manually).
I know that you are the owner of Starling Framework. Can I use **Starling Stage's drawToBitmapData** to solve this problem (render my 3D objects created in Animate)? (Of course, I just use the "drawToBitmapData " of this framework but not re-create everything in my app by this framework).
I am thinking a feature in AIR, a library base on a primitive feature of AIR's source code (that can render the graphic FROM the stage better).
2. You said that:
`OBS Studio is an open source program licensed under the [GPLv2](https://github.com/obsproject/obs-studio/blob/master/COPYING). That means that if you create a plugin for OBS Studio or a modified version of OBS Studio and distribute it in binary form, you must give users a way to access the source code of your binary.`
But why I "must give users a way to access the source code" of my plug-in? I think I can write it and use for myself. Can you explain more about this?

**Reply by ajwfrost** [reply]
Yes I can see that drawing the stage to bitmap data and sending this to FFMPEG would be a bit slow!

I'm not the owner of Starling, but yes you could use drawToBitmapData to render 3D contents into a bitmap data object - but this is a very slow operation as it pulls everything out of the 3D pipeline which causes that to stall.

Creating a way to expose the stage contents from AIR isn't necessarily straightforward especially as it depends on the render mode and GPU mechanisms.. if you're wanting to move pixel data around, then rendering as 'cpu' (or even 'direct') would be okay (we may need to update the implementation somewhat to optimize it though, and that doesn't count Stage3D-rendered content). It would be better to use some Direct3D sharing mechanism hence 'direct' or 'gpu' modes perhaps, which would work for the 3D content but would definitely need more integration work.

In terms of the OBS licensing, it's only if you give your plug-in to anyone, you must also give them your source code. That's a condition of using their software, due to the GPL license conditions. It's a license we stay well away from because it has this contamination clause that means if you integrate your software with a GPL component, your software then falls under GPL itself. If you're just doing it for yourself then I think it's irrelevant, it's only if you give anyone a copy.
